---
title: "TCGA - Expression and Survival"
output: html_document
date: "2024-08-14"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load packages 
```{r}
library(TCGAbiolinks)
library(SummarizedExperiment)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggpubr)
library(ggforce)
library(ggExtra)
library(ggtext)
library(ggfortify)
library(grid)
library(gridExtra)
library(ggridges)
library(ggbeeswarm)
library(ggsurvfit)
library(survminer)
library(survival)
library(readxl)
library(stringr)
```

## Establishing ggplot theme arguments
```{r}
graph_theme <- theme(
  plot.margin = margin(b = 0.2, l = 0.2, t = 0.2, r = 0.2, unit = "lines"),
  text = element_text(size = 1, family = "Arial"),
  panel.spacing.y = unit(0.1, "lines"),
  panel.spacing.x = unit(0.1, "lines"),
  panel.border = element_rect(colour = "#000000", fill = NA, linewidth = 0.5),
  panel.background = element_blank(),
  panel.grid.major.y = element_blank(),
  panel.grid.minor.y = element_blank(),
  panel.grid.major.x = element_blank(),
  panel.grid.minor.x = element_blank(),
  strip.background = element_blank(),
  strip.text.x = element_text(colour = "#000000", size = 6, margin = margin(b = 2, t = 2), vjust = 0, family = "Arial"),
  strip.text.y = element_text(colour = "#000000", size = 6, margin = margin(l = 2), family = "Arial"),
  strip.clip = "off",
  axis.ticks.x = element_line(colour = "#000000", linewidth = 0.25),
  axis.ticks.y = element_line(colour = "#000000", linewidth = 0.25),
  axis.ticks.length = unit(0.1, "lines"),
  axis.title.x = element_text(colour = "#000000", size = 6, margin = margin(t = 3), family = "Arial"),
  axis.title.y = element_text(colour = "#000000", size = 6, margin = margin(r = 2), family = "Arial"),
  axis.text.x = element_text(colour = "#000000", size = 6, margin = margin(t = 2), family = "Arial"),
  axis.text.y = element_text(colour = "#000000", size = 6, margin = margin(r = 2), family = "Arial"),
  plot.title = element_blank(),
  legend.key = element_rect(fill = "#FFFFFF"),
  legend.position = "none",
  legend.justification = "left",
  legend.key.width = unit(2, 'pt'),
  legend.key.height = unit(2, 'pt'),
  legend.text = element_text(colour = "#000000", size = 6, margin = margin(r = 1), family = "Arial"),
  legend.margin = margin(r = 0, b = 0, l = 0, t = 0),
  legend.title = element_blank())
```

## Establishing colour pallette
```{r}
colour_pallette <- c("#D4D3CF", "#DC6B83", "#75B1CE", "#D8C367", "#526C94", "#000000", "#ccdba2", "#889466", "#c6b2d1", "#654875")
```

## Set working directory
```{r}
working_directory <- "/Users/rz73/Documents/Lab-book/The Cancer Genome Atlas (TCGA) Analysis"
```

## Downloading TCGA gene expression data
https://www.bioconductor.org/packages/devel/bioc/vignettes/TCGAbiolinks/inst/doc/query.html
```{r}
setwd(working_directory)
path_GDCdata <- paste(getwd(), "/GDCdata/", sep = "")

list_projects <- c("TCGA-ACC", "TCGA-BLCA", "TCGA-BRCA",
                   "TCGA-CESC", "TCGA-CHOL", "TCGA-COAD",
                   "TCGA-DLBC", "TCGA-ESCA", "TCGA-GBM",
                   "TCGA-HNSC", "TCGA-KICH", "TCGA-KIRC",
                   "TCGA-KIRP", "TCGA-LAML", "TCGA-LGG",
                   "TCGA-LIHC", "TCGA-LUAD", "TCGA-LUSC",
                   "TCGA-MESO", "TCGA-OV", "TCGA-PAAD",
                   "TCGA-PCPG", "TCGA-PRAD", "TCGA-READ",
                   "TCGA-SARC", "TCGA-SKCM", "TCGA-STAD",
                   "TCGA-TGCT", "TCGA-THCA", "TCGA-THYM",
                   "TCGA-UCEC", "TCGA-UCS", "TCGA-UVM")

for (p in list_projects) {
  
  print(paste("Downloading data for project", p, sep = " "))
  query_TCGA = GDCquery(
    project = p,
    data.type = "Gene Expression Quantification",
    data.category = "Transcriptome Profiling",
    experimental.strategy = "RNA-Seq",
    workflow.type = "STAR - Counts")
  
  GDCdownload(query = query_TCGA, directory = path_GDCdata)
}
```

## Associating RNA expression data with clinical survival data for the gene of interest
```{r}
### Setting the working directory
setwd(working_directory)
### Establishing the gene of interest (GOI)
GOI <- "PPP2R2A"
TCGA_genes <- read_excel(paste(getwd(), "/data/", "TCGA_gene_list.xlsx", sep = ""))
TCGA_GOI <- filter(TCGA_genes, gene_name == GOI)
### Extracting the ensembl id of the gene of interest (GOI)
GOI_ID <- TCGA_GOI$gene_id
### Creating the directory to store the output data
dir.create(paste(getwd(), "/outputs/", GOI, sep = ""), recursive = TRUE)

### Clinical studies of interest
list_projects <- c("TCGA-ACC", "TCGA-BLCA", "TCGA-BRCA",
                   "TCGA-CESC", "TCGA-CHOL", "TCGA-COAD",
                   "TCGA-DLBC", "TCGA-ESCA", "TCGA-GBM",
                   "TCGA-HNSC", "TCGA-KICH", "TCGA-KIRC",
                   "TCGA-KIRP", "TCGA-LAML", "TCGA-LGG",
                   "TCGA-LIHC", "TCGA-LUAD", "TCGA-LUSC",
                   "TCGA-MESO", "TCGA-OV", "TCGA-PAAD",
                   "TCGA-PCPG", "TCGA-PRAD", "TCGA-READ",
                   "TCGA-SARC", "TCGA-SKCM", "TCGA-STAD",
                   "TCGA-TGCT", "TCGA-THCA", "TCGA-THYM",
                   "TCGA-UCEC", "TCGA-UCS", "TCGA-UVM")

for (p in list_projects) {
  
  tmp.project <- strsplit(p, split = "-")[[1]][2]
  
  if (paste(tmp.project, ".csv", sep = "") %in% list.files(paste(getwd(), "/", GOI_ID, "/", sep = ""))) {
    
    print(paste("Data for project", p, "already downloaded", sep = " "))
    
  } else {
    
    ### TCGA query
    query_TCGA = GDCquery(
    project = p,
    data.type = "Gene Expression Quantification",
    data.category = "Transcriptome Profiling",
    experimental.strategy = "RNA-Seq",
    workflow.type = "STAR - Counts")
    TCGA_data <- GDCprepare(query_TCGA)
    
    ### Extracting clinical information
    data_clinical <- data.frame(colData(TCGA_data))
    
    if (!("ajcc_pathologic_stage" %in% colnames(data_clinical))) {
      
      data_clinical <- mutate(data_clinical, ajcc_pathologic_stage = "not_reported")
    }
    
    data_clinical <- dplyr::select(data_clinical, barcode, gender, sample_type, ajcc_pathologic_stage, vital_status, days_to_death, days_to_last_follow_up)
    
    ### Extracting fragments per kilobase of transcript per million mapped reads (FPKM) values for GOI
    data_GOI_expression <- data.frame(assays(TCGA_data)[["fpkm_unstrand"]])[GOI_ID, ]
    data_GOI_expression <- mutate(data_GOI_expression, ensembl_id = rownames(data_GOI_expression)) %>%
      gather(key = "barcode", value = "fpkm", 1:(length(.) - 1)) %>%
      mutate(barcode = gsub("\\.", "-", barcode))
    
    ### Merging expression (FPKM) and clinical data
    data_merged <- merge(data_clinical, data_GOI_expression, by = "barcode") %>%
      mutate(project = strsplit(p, split = "-")[[1]][2]) %>%
      dplyr::select(project, barcode, gender, sample_type, ajcc_pathologic_stage, vital_status, days_to_death, days_to_last_follow_up, ensembl_id, fpkm)
    
    ### Saving data
    write.csv(data_merged, paste(getwd(), "/outputs/", GOI, "/", strsplit(p, split = "-")[[1]][2], ".csv", sep = ""), row.names = FALSE)
    } 
}
```

## Processing RNA expression and survival data
```{r}
### Setting the working directory
setwd(working_directory)
### Establishing GOI
GOI <- "PPP2R2A"

### Importing raw survival data
data_survival <- data.frame() %>%
  (function (data) {
    
    for (file in list.files(paste(getwd(), "/outputs/", GOI, "/" , sep = ""))) {
      tmp.data <- read.csv(paste(getwd(), "/outputs/", GOI, "/" , file, sep = ""), sep = ",")
      data <- bind_rows(data, tmp.data)
    }
    return(data)
  }) %>%
  mutate(sample_type = case_when(sample_type == "Primary Blood Derived Cancer - Peripheral Blood" ~ "Primary Tumor",
                                 .default = sample_type),
         deceased = case_when(vital_status == "Dead" ~ TRUE,
                              vital_status == "Alive" ~ FALSE),
         survival = case_when(vital_status == "Dead" ~ days_to_death,
                              vital_status == "Alive" ~ days_to_last_follow_up))

### Processing raw survival data
data_survival_processed <- data_survival %>%
  filter(sample_type %in% c("Primary Tumor", "Solid Tissue Normal")) %>%
  mutate(sample_type = factor(sample_type, levels = c("Solid Tissue Normal", "Primary Tumor"), labels = c("N", "T"))) %>%
  group_by(project, sample_type) %>%
  mutate(expression = case_when(fpkm <= quantile(fpkm, 0.5) ~ "Low", fpkm > quantile(fpkm, 0.5) ~ "High")) %>%
  ungroup()

data_survival_analysis <- data.frame()

## Constructing Kaplan-Meier survival dataset
for (p in unique(data_survival_processed$project)) {
  
  data_survival_subset <- data_survival_processed %>%
    filter(sample_type == "T", project == p, !is.na(survival))
  
  Kaplan_Meier_fit <- survfit(Surv(survival, deceased) ~ expression, data = data_survival_subset)
  Kaplan_Meier_stats <- survdiff(Surv(survival, deceased) ~ expression, data = data_survival_subset)
  
  # rmean represents Restricted Mean Survival Time (RMST)
  # rmean = "common" parameter ensures that RMST is calculated over the same time horizon for all groups being compared
  
  Kaplan_Meier_summary <- tibble::rownames_to_column(data.frame(summary(Kaplan_Meier_fit, rmean = "common")$table)) %>%
    mutate(project = p) %>%
    separate(rowname, into = c("strata", "expression")) %>%
    dplyr::select(project, expression, rmean) %>%
    spread(key = expression, value = rmean) %>%
    mutate(rmean_FC = (High - Low) / (High + Low), p = Kaplan_Meier_stats$pvalue)
  
  data_survival_analysis <- bind_rows(data_survival_analysis, Kaplan_Meier_summary)
}

head(data_survival_processed)
head(data_survival)
head(data_survival_analysis)
```
## Plotting cancer-specific expression profiles for GOI
```{r}
### Setting the working directory
setwd(working_directory)

data_graph_expression_profiles_summary <- data_survival_processed %>%
  reframe(fpkm_median = median(fpkm),
          .by = c(project, sample_type))

data_expression_profiles_stats <- data_survival_processed %>%
  dplyr::select(project, barcode, sample_type, fpkm) %>%
  spread(key = sample_type, value = fpkm) %>%
  group_by(project) %>%
  group_split() %>%
  lapply((function (data) {
    
    tmp.N <- data[["N"]][!(is.na(data[["N"]]))]
    tmp.T <- data[["T"]][!(is.na(data[["T"]]))]
    
    if (length(tmp.N) < 3 | length(tmp.T) < 3) {
      
      tmp.output <- data.frame(project = unique(data$project), p = NA)
    } else {
      
      tmp.p <- t.test(tmp.N, tmp.T, alternative = "two.sided", var.equal = FALSE)$p.value
      tmp.output <- data.frame(project = unique(data$project), p = tmp.p, mean_diff = mean(tmp.T) - mean(tmp.N))
    }
    
    return(tmp.output)})) %>%
  
  bind_rows() %>%
  mutate(significance = case_when(p < 0.00001 ~ "*****",
                                  p < 0.0001 ~ "****",
                                  p < 0.001 ~ "***",
                                  p < 0.01 ~ "**",
                                  p < 0.05 ~ "*",
                                  p >= 0.05 ~ "ns",
                                  is.na(p) ~ "nd"))

graph_expression_profiles <- ggplot() +
  
  geom_quasirandom(data = data_survival_processed, aes(x = sample_type, y = fpkm, colour = sample_type),
                   size = 0.75, alpha = 0.25, shape = 16) +
  geom_point(data = data_graph_expression_profiles_summary, aes(x = sample_type, y = fpkm_median),
             size = 0.75, stroke = 0.35, shape = 21, colour = "#000000", fill = "#FFFFFF") +
  
  geom_text(data = filter(data_expression_profiles_stats, significance %in% c("nd", "ns")),
            aes(x = 1.5, y = 1.1 * max(data_survival_processed$fpkm), label = significance),
            size = 6/.pt, colour = colour_pallette[1], family = "Arial") +
  
  geom_text(data = filter(data_expression_profiles_stats, !(significance %in% c("nd", "ns")), mean_diff < 0),
            aes(x = 1.5, y = 1.1 * max(data_survival_processed$fpkm), label = significance),
            size = 6/.pt, colour = colour_pallette[3], family = "Arial") +
  
  geom_text(data = filter(data_expression_profiles_stats, !(significance %in% c("nd", "ns")), mean_diff > 0),
            aes(x = 1.5, y = 1.1 * max(data_survival_processed$fpkm), label = significance),
            size = 6/.pt, colour = colour_pallette[2], family = "Arial") +
  
  scale_colour_manual(values = c(colour_pallette[3], colour_pallette[2])) +
  scale_y_continuous(limits = c(0, 1.2 * max(data_survival_processed$fpkm))) +
  facet_wrap(~ project, nrow = 3) +
  labs(x = "Tissue (N - Normal, T - Tumor)", y = paste(GOI, " expression (FPKM)", sep = "")) +
  graph_theme

### Creating the directory to store generated graphs
if (!dir.exists(paste(getwd(), "/outputs/", "graphs/", sep = ""))) {
  dir.create(paste(getwd(), "/outputs/", "graphs/", sep = ""), recursive = TRUE)
  cat("Directory created successfully.\n")
} else {
  cat("Directory already exists.\n")
}

ggsave(paste(getwd(), "/outputs/", "graphs/", GOI, "_expression.pdf", sep = ""), graph_expression_profiles,
       width = 85, height = 50 , units = "mm", device = cairo_pdf)
```

```{r}
### Setting the working directory
setwd(working_directory)

data_graph_survival <- data_survival_processed %>%
  
  group_by(project) %>%
  group_split() %>%
  lapply((function (data) {
    
    fit = survfit(Surv(survival, deceased) ~ expression, data = data)
    fit_summary <- summary(fit, times = seq(0, 6000, 1), extend = T)
    
    tmp.data <- tibble("time" = fit_summary$time / 365,
                       "n.risk" = fit_summary$n.risk,
                       "n.event" = fit_summary$n.event,
                       "n.censor" = fit_summary$n.censor,
                       "estimate" = fit_summary$surv,
                       "std.error" = fit_summary$std.err,
                       "strata" = fit_summary$strata) %>%
      mutate(project = unique(data$project))
    return(tmp.data)})) %>%
  bind_rows() %>%
  mutate(expression = str_split_fixed(as.character(strata), pattern = "=", n = 2)[, 2]) %>%
  mutate(expression = factor(expression, levels = c("Low", "High")))

graph_survival <- ggplot() +
    
  geom_ribbon(data = data_graph_survival,
              aes(x = time, ymin = estimate - std.error, ymax = estimate + std.error, fill  = expression), alpha = 0.15) +
  
  geom_step(data = data_graph_survival,
            aes(x = time, y = estimate, color  = expression),
            linewidth = 0.3) +

  scale_fill_manual(values = c(colour_pallette[5], colour_pallette[2])) +
  scale_colour_manual(values = c(colour_pallette[5], colour_pallette[2])) +
  facet_wrap(~ project, nrow = 3) +
  scale_x_continuous(expand = c(0, 0), breaks = c(2, 8, 14)) +
  scale_y_continuous() +
  labs(x = "Years after diagnosis", y = "Probability of survival", colour = paste(GOI, "expression", sep = " ")) +
  graph_theme +
  theme(
    legend.position = "top",
    legend.justification = "left",
    legend.background = element_blank(),
    legend.box = element_blank(),
    legend.key=element_blank(),
    legend.key.width = unit(10, 'pt'),
    legend.key.height = unit(6, 'pt'),
    legend.text = element_text(colour = "#000000", size = 6, margin = margin()),
    legend.margin = margin(b = -8),
    legend.spacing.x = unit(1, "pt"),
    legend.title = element_text(colour = "#000000", size = 6)) +
  
  guides(fill = FALSE)

### Creating the directory to store generated graphs
if (!dir.exists(paste(getwd(), "/outputs/", "graphs/", sep = ""))) {
  dir.create(paste(getwd(), "/outputs/", "graphs/", sep = ""), recursive = TRUE)
  cat("Directory created successfully.\n")
} else {
  cat("Directory already exists.\n")
}

ggsave(paste(getwd(), "/outputs/", "graphs/", GOI, "_survival.pdf", sep = ""), graph_survival,
       width = 85, height = 50 , units = "mm", device = cairo_pdf)
```
